arangod/Aql/Executor/EnumerateListExecutor.cpp

```diff
@@ -364,5 +381,227 @@ AqlValue EnumerateListExecutor::getAqlValue(AqlValue const& inVarReg,
+    // we can only skip the max amount of values we've in our array
+    skipped = length - pos;
+  }
+  return skipped;
+}
+
+std::tuple<ExecutorState, FilterStats, AqlCall>
+EnumerateListObjectExecutor::produceRows(AqlItemBlockInputRange& inputRange,
+                                         OutputAqlItemRow& output) {
+  FilterStats stats{};
+
+  AqlCall upstreamCall{};
+  upstreamCall.fullCount = output.getClientCall().fullCount;
+
+  while (inputRange.hasDataRow() && !output.isFull()) {
+    if (!_objIterator.valid()) {
+      // we reached either the end of an array
+      // or are in our first loop iteration
+      initializeNewRow(inputRange);
+      continue;
+    }
+
+    TRI_ASSERT(_objIterator.valid());
+
+    if (!processElement(output)) {
+      // item was filtered out
+      stats.incrFiltered();
+    }
+    _objIterator.next();
+
+    _killCheckCounter = (_killCheckCounter + 1) % 1024;
+    if (ADB_UNLIKELY(_killCheckCounter == 0 && _infos.getQuery().killed())) {
+      THROW_ARANGO_EXCEPTION(TRI_ERROR_QUERY_KILLED);
+    }
+  }
+
+  if (!_objIterator.valid()) {
+    // we reached either the end of an array
+    // or are in our first loop iteration
+    initializeNewRow(inputRange);
+  }
+
+  return {inputRange.upstreamState(), stats, upstreamCall};
+}
+
+std::tuple<ExecutorState, FilterStats, size_t, AqlCall>
+EnumerateListObjectExecutor::skipRowsRange(AqlItemBlockInputRange& inputRange,
+                                           AqlCall& call) {
+  FilterStats stats{};
+
+  while (inputRange.hasDataRow() && call.needSkipMore()) {
+    if (!_objIterator.valid()) {
+      // we reached either the end of an array
+      // or are in our first loop iteration
+      initializeNewRow(inputRange);
+      continue;
+    }
+
+    TRI_ASSERT(_objIterator.valid());
+
+    if (_infos.hasFilter()) {
+      // we have a filter. we need to apply the filter for each
+      // document first, and only count those as skipped that
+      // matched the filter.
+
+      // TODO: Implement filter
+
+      // AqlValue const& inputList =
+      // _currentRow.getValue(_infos.getInputRegister());
+      // bool mustDestroy;
+      // AqlValue innerValue =
+      //     getAqlValue(inputList, _inputArrayPosition, mustDestroy);
+      // AqlValueGuard guard(innerValue, mustDestroy);
+
+      // if (checkFilter(innerValue)) {
+      //   call.didSkip(1);
+      // } else {
+      //   stats.incrFiltered();
+      // }
+
+      // always advance input position
+      _objIterator.next();
+    } else {
+      // no filter. we can skip many documents at once
+      auto const skip = std::invoke([&] {
+        // if offset is > 0, we're in offset skip phase
+        if (call.getOffset() > 0) {
+          // we still need to skip offset entries
+          return call.getOffset();
+        } else {
+          TRI_ASSERT(call.needsFullCount());
+          // fullCount phase - skippen bis zum ende
+          return _objIterator.index();
```

Comment:
Not sure if this is correct, I would expect something like `size - index` to compute how many elements there are left to skip.